src/mail-app/mail/model/MailListModel.ts

```diff
@@ -0,0 +1,314 @@
+import { ListFilter, ListModel } from "../../../common/misc/ListModel"
+import { Mail, MailFolder, MailFolderTypeRef, MailSetEntry, MailSetEntryTypeRef, MailTypeRef } from "../../../common/api/entities/tutanota/TypeRefs"
+import {
+	CUSTOM_MAX_ID,
+	deconstructMailSetEntryId,
+	elementIdPart,
+	firstBiggerThanSecond,
+	getElementId,
+	isSameId,
+	listIdPart,
+} from "../../../common/api/common/utils/EntityUtils"
+import { EntityClient } from "../../../common/api/common/EntityClient"
+import { ConversationPrefProvider } from "../view/ConversationViewModel"
+import { assertMainOrNode } from "../../../common/api/common/Env"
+import { assertNotNull } from "@tutao/tutanota-utils"
+import { ListLoadingState, ListState } from "../../../common/gui/base/List"
+import Stream from "mithril/stream"
+import { EntityUpdateData, isUpdateForTypeRef } from "../../../common/api/common/utils/EntityUpdateUtils"
+import { OperationType } from "../../../common/api/common/TutanotaConstants"
+
+assertMainOrNode()
+
+type LoadedMail = {
+	mail: Mail
+	mailSetEntry: MailSetEntry
+}
+
+/** sort mail set mails in descending order (**reversed**: newest to oldest) according to their receivedDate, not their elementId */
+export function sortCompareMailSetMailsReversed(firstMail: Mail, secondMail: Mail): number {
+	const timestampDifference = secondMail.receivedDate.getTime() - firstMail.receivedDate.getTime()
+	if (timestampDifference !== 0) {
+		return timestampDifference
+	}
+
+	const firstMailElementId = getElementId(firstMail)
+	const secondMailElementId = getElementId(secondMail)
+	if (firstBiggerThanSecond(firstMailElementId, secondMailElementId)) {
+		return -1
+	} else if (firstBiggerThanSecond(secondMailElementId, firstMailElementId)) {
+		return 1
+	} else {
+		return 0
+	}
+}
+
+/**
+ * Handles fetching and resolving mail set entries into mails as well as handling sorting.
+ */
+export class MailListModel {
+	// Id = MailSetEntry element id
+	private readonly listModel: ListModel<LoadedMail, Id>
+
+	// keep a reverse map for going from Mail element id -> LoadedMail
+	private readonly mailMap: Map<Id, LoadedMail> = new Map()
+
+	constructor(
+		private readonly mailSet: MailFolder,
+		private readonly conversationPrefProvider: ConversationPrefProvider,
+		private readonly entityClient: EntityClient,
+	) {
+		this.listModel = new ListModel({
+			fetch: async (lastFetchedItem, count) => {
+				const lastFetchedId = lastFetchedItem?.mailSetEntry?._id ?? [mailSet.entries, CUSTOM_MAX_ID]
+				const items = await this.loadMails(lastFetchedId, count)
+				return {
+					items,
+					complete: items.length < count,
```

Comment:
`complete` condition was different before because of inbox rules, should we take that into account?